streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -255,53 +249,53 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
-                      .flatMap(_ => f(a))
-                      .foreach(b => out.offer(Take.Value(b)).unit)
-                      .foldCauseM(
-                        cause => out.offer(Take.Fail(cause)) *> innerFailure.fail(cause),
-                        _ => ZIO.unit
-                      )
-                    _ <- (innerStream race interruptInners.await).fork
-                    // Make sure that the current inner stream has actually succeeded in acquiring
-                    // a permit before continuing. Otherwise, two bad things happen:
-                    // - we might needlessly fork inner streams without available permits
-                    // - worse, we could reach the end of the stream and acquire the permits ourselves
-                    //   before the inners had a chance to start
-                    _ <- latch.await
-                  } yield ()
-                }.foldCauseM(
-                    cause =>
-                      (interruptInners.succeed(()) *> permits.acquireN(n) *> out
-                        .offer(Take.Fail(cause))).unit.toManaged_,
-                    _ =>
-                      innerFailure.await
-                        .raceWith(permits.acquireN(n))(
-                          // One of the inner fibers failed. It already enqueued its failure, so we
-                          // just need to interrupt the inner fibers and acquire all permits to await
-                          // their termination
-                          leftDone = (_, permitAcquisition) => interruptInners.succeed(()) *> permitAcquisition.join,
-                          // All fibers completed successfully, so we just need to signal that
-                          // we're done.
-                          rightDone = (_, failureAwait) => out.offer(Take.End) *> failureAwait.interrupt
-                        )
-                        .toManaged_
-                  )
-                  // Forking the ZManaged value means that on the finalization step, the fiber would
-                  // be interrupted, so we don't have to worry that the driver would keep spawning
-                  // more inner fibers when the consuming stream is short circuited
-                  .fork
+            driver <- self.foreachManaged { a =>
+                       for {
+                         latch <- Promise.make[Nothing, Unit]
+                         innerStream = Stream
+                           .bracket(permits.acquire *> latch.succeed(()))(_ => permits.release)
+                           .flatMap(_ => f(a))
+                           .foreach(b => out.offer(Take.Value(b)).unit)
+                           .foldCauseM(
+                             cause => out.offer(Take.Fail(cause)) *> innerFailure.fail(cause),
+                             _ => ZIO.unit
+                           )
+                         _ <- (innerStream race interruptInners.await).fork
```

Comment:
isn't there a chance here that `interruptInners.await` wins the race before `innerStream` has had the chance to eval `latch.succeed`?

Reply:
Erm, no.

Reply:
I think that's ok because only the short-circuiting/interruption finalizer would be the one triggering `interruptInners` in that case.